### PR TITLE
[FIX] http_routing: remove trailing `/`

### PR DESCRIPTION
**Pull request by ngochuy97hp**
Currently, the character `/` appears at the end of the url of the
language dropdown on the home page e.g. /en/, /fr/. This will cause one
more redirect when performing the language change.

Indeed, this error is only encountered with path = /
Please try the following `url_lang('/', 'en_US')` => /en/

After this fix, the trailing `/` will be removed when using the func
`url_lang` e.g. `url_lang('', 'en_US')` => /en, this means that the
language switching links on the homepage no longer redirect redundantly
again.

Screenshot from 2022-11-19 11-08-28

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ngochuy97hp** [comment]
@rdeodoo Hello @rdeodoo :wave:, could you take a look at this PR?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Hello, please see https://github.com/odoo/odoo/commit/5945e8e4e353ff45657e59b600464254839bd5c9
This would be creating some other issues (for a not really critical case)

**Comment by ngochuy97hp** [comment]
> Hello, please see [5945e8e](https://github.com/odoo/odoo/commit/5945e8e4e353ff45657e59b600464254839bd5c9) This would be creating some other issues (for a not really critical case)

Thank you, what are those cases? :astonished:

**Comment by rdeodoo** [comment]
I believe it would be breaking the routes defined with a trailing slash, as `@route('/myroute/')`.
We removed those from our codebase (see https://github.com/odoo/odoo/commit/fee01131d495c61d98cc7fb31c59f1541cbbc01e)
As it was a late change before the 15.0 release, we [didn't officially stop supporting](https://github.com/odoo/odoo/pull/80109#issuecomment-975505055) trailing slash, so we couldn't keep this commit that would break such routes.

**Comment by ngochuy97hp** [comment]
@rdeodoo Thank you, I tried it, it will take one more redirect for links with trailing slash. Not bad either :joy:

**Comment by rdeodoo** [comment]
It was not about a redirect issue, it was really breaking the dispatcher (not landing where it should -> 404) in case of multi lang

**Comment by ngochuy97hp** [comment]
[test-route.webm](https://user-images.githubusercontent.com/66666640/206337828-3d052b7f-1612-431c-82e8-ea1b7c53d0fb.webm)

Hello @rdeodoo. I've tested it, but I don't get that error, I don't know if that's enough?

**Comment by rdeodoo** [comment]
@JKE-be Maybe you remember what flow (and what error) you were speaking about here: https://github.com/odoo/odoo/commit/5945e8e4e353ff45657e59b600464254839bd5c9
> Unfortunately, it will break some route with trailing / in case of multi lang.

**Comment by rdeodoo** [comment]
Ok I finally found what was the bug described in the revert commit:
- Add a trailing slash to a route (for testing purpose, just add on to the blog post controller)
- Add second lang, like french
- As public user, go to the blog without the trailing slash (with or without the lang, doesn't matter): `http://127.0.0.1:8069/blog/voyager-1/sierra-tarahumara-1` (or with lang `http://127.0.0.1:8069/fr_BE/blog/voyager-1/sierra-tarahumara-1`

You will enter an infinite loop:
image

Note that the revert commit actually reverted the same code as this PR but which was not indented the same way, yours is in the else block.

All in all, I'd not merge this kind of change in stable, the risk reward is really bad.
If you want to retarget master and do another round of test, we can probably make it work.

**Comment by ngochuy97hp** [comment]
Thanks. Actually, I got a redirect loop error if I put the code outside, otherwise it's ok like this PR.

**Comment by rdeodoo** [comment]
I am ok with reviewing and merging the change, but in master as I said.
It won't bring a lot of added value (low reward), for a not -so-low risk, just for a perf imp in the language switcher.
I'll also add a test with your commit (once ready to be merged) to prevent the infinite loop should someone one day change it and introduce the problematic code we discussed.

**Comment by ngochuy97hp** [comment]
Okay @rdeodoo, I changed the target to master

**Comment by rdeodoo** [comment]
I am pretty sure there will be cases where we will end up in the infinite redirect. I'd like to properly stress test it before merge to be sure, but:
- I found https://github.com/odoo/odoo/pull/108478 that actually prevent me to correctly test it
- I'll be on leave until 2023

So I'll have a look in a few weeks actually, thank you for your contribution and patience.

**Review by rdeodoo — APPROVED** [review]
@robodoo r+

Thank you for the changes.
Note that the infinite loop (from the reverted commit we discussed / if unindented) was probably coming from the use of `url_for()` in dispatcher before httpocalypse. It's no more part of the dispatcher now, so probably was fine to implement the fix without the extra indent (basically restoring the reverted commit).
See https://github.com/odoo/odoo/blame/f662051e385201cd658dab18b5880a9db4db1355/addons/http_routing/models/ir_http.py#L461

**Comment by JKE-be** [comment]
broken case with this:

`url_for('/forum/') + slug(record)`

Probably a wrong use case, but it exists...

https://github.com/odoo/odoo/pull/120438

**Pull request closed without merging**
